Final/RFC00XX-Custom-Remote-Connections.md

```diff
@@ -0,0 +1,281 @@
+PS C:\> $runspace
+
+ Id Name            ComputerName    Type          State         Availability
+ -- ----            ------------    ----          -----         ------------
+  2 Runspace2       remote-1        Remote        Opened        Available
+```
+
+### Endpoints
+
+A remote connection must include a PowerShell session on the target machine, that handles remoting protocol messages from the client.
+This is called the remoting endpoint.
+An endpoint typically involves a service that listens for client connection requests, and securely creates the PowerShell session under the correct account and with correct permissions.
+For example, both WSMan and SSH have services that do this.
+Endpoints such as these are outside the scope of this feature proposal, and it is assumed that any custom connection scheme will include some sort of endpoint support.  
+
+PowerShell currently supports a 'server' running mode, where remoting protocol messages are channeled through the process stdIn/stdOut data streams.
+Any custom endpoint solution will leverage this to establish a PowerShell session and handle remoting protocol messages.
+Currently, this is the only way to run PowerShell as an endpoint server, and should be sufficient for any custom solution.
+Also this server mode is available on all supported PowerShell versions (v5.1 - v7.1+).
+
+#### Example - PowerShell in server mode
+
+This example illustrates how PowerShell can be run in server mode.
+
+```powershell
+# Create PowerShell endpoint session (with default shell configuration)
+# Protocol handles messages through the child process stdIn/stdOut process pipes
+$proc = Start-Process -FilePath pwsh.exe -ArgumentList '-servermode -nologo -noprofile' -PassThru
+
+# Pass child process object to custom connection endpoint handler that reads/writes connection streams from stdIn/stdOut pipes
+[MyCustomConnection.CustomConnectionEndpoint]::SetEndPointProcess($proc)
+```
+
+### Implementation Example
+
+This is a simple example of how to create a custom ConnectionInfo and Transport implementation
+
+```csharp
+namespace MyCustomConnection
+{
+    // Derived custom connection information class
+    public sealed class CustomConnectionInfo : RunspaceConnectionInfo
+    {
+        // New multi-factor authentication switch for connection.
+        // Otherwise use existing 'ComputerName' and 'Credential' properties from derived class.
+        public SwitchParameter MFA { get; private set; }
+
+        // Message reader and writer.
+        internal StreamReader messageReader { get; private set; }
+        internal StreamWriter messageWriter { get; private set; }
+
+        // Override and implement this method that creates the custom client session transport with
+        // bound live connection data streams.
+        // This method is called from within the internal PowerShell remoting implementation.
+        public override BaseClientSessionTransportManager CreateClientSessionTransportManager(
```

Comment:
I don't know how far in scope the RFC should be but it might be good to document what each of these options are for.

Reply:
I don't think it is needed in this proposal, but if we move forward all of this will be documented.